### PR TITLE
Updated the explanation of ONBUILD in best-practices doc

### PR DESCRIPTION
**Pull request by TonnyXu**
## Reason

The original explanation on `ONBUILD` instruction is not easy to understand.
## Changes

Added an explanation for `ONBUILD` based on parent/child Dockerfile context

### CODE REVIEW

**Comment by TonnyXu** [comment]
Signed.

File: docs/sources/articles/dockerfile_best-practices.md [hunk 1]
```diff
@@ -398,6 +398,14 @@ troubleshoot, and maintain.
 
 ### [`ONBUILD`](https://docs.docker.com/reference/builder/#onbuild)
 
+An `ONBUILD` command in a Dockerfile will not be executed when this
```
Comment (moxiegirl) [inline docs/sources/articles/dockerfile_best-practices.md]: Stating things in the negative and in the future can be hard to parse.  Instead, try this for easier reading.  You'll find the markdown in this gist:

https://gist.github.com/moxiegirl/b93d75a32a9c82cd03f3


-----
An `ONBUILD` command executes after the current Dockerfile build completes;
`ONBUILD` executes for any child image derived `FROM` the current image. Think
of the `ONBUILD` command as an instruction that parent Dockerfile gives to the child
Dockerfile.

A Docker build executes `ONBUILD` commands before any command in a child
Dockerfile.

`ONBUILD` is useful for images that are built `FROM` a given image. For example,
you would use `ONBUILD` for a language stack image that builds arbitrary user
software written in that language within the `Dockerfile`, as you can see in
[Ruby’s `ONBUILD`
variants](https://github.com/docker-library/ruby/blob/master/2.1/onbuild/
Dockerfile).

Tag images built from `ONBUILD` separately from the parent, for example:
`ruby:1.9-onbuild` or `ruby:2.0-onbuild`.

Be careful when putting `ADD` or `COPY` in `ONBUILD`. The “onbuild” image can
fail catastrophically if the new build's context is missing a resource being
added or copied. Adding a separate tag, as recommended above, may help mitigate
this by allowing the `Dockerfile` author to make a choice.

File: docs/sources/articles/dockerfile_best-practices.md [hunk 2]
```diff
@@ -398,7 +398,15 @@ troubleshoot, and maintain.
 
 ### [`ONBUILD`](https://docs.docker.com/reference/builder/#onbuild)
 
-`ONBUILD` is only useful for images that are going to be built `FROM` a given
```
Comment (TonnyXu) [inline docs/sources/articles/dockerfile_best-practices.md]: @moxiegirl Great! thanks for your great advice!

File: docs/sources/articles/dockerfile_best-practices.md [hunk 3]
```diff
@@ -398,7 +398,15 @@ troubleshoot, and maintain.
...
+An `ONBUILD` command executes after the current Dockerfile build completes.
```
Comment (jamtur01) [inline docs/sources/articles/dockerfile_best-practices.md]: \`Dockerfile\`

File: docs/sources/articles/dockerfile_best-practices.md [hunk 4]
```diff
@@ -398,7 +398,15 @@ troubleshoot, and maintain.
...
+`ONBUILD` executes in any child image derived `FROM` the current image.  Think
+of the `ONBUILD` command as an instruction the parent Dockerfile gives
```
Comment (jamtur01) [inline docs/sources/articles/dockerfile_best-practices.md]: As above.

**Comment by jamtur01** [comment]
LGTM ping @moxiegirl @fredlf @SvenDowideit

**Comment by moxiegirl** [comment]
Thank you for the contribution. LGTM

**Comment by moxiegirl** [comment]
Tonny,

Can you rebase and squash your commits?   Here is a procedure that shows you how:

http://docs.docker.com/project/create-pr/#rebase-your-branch

**Comment by fredlf** [comment]
LGTM, thanks Tonny.

**Comment by TonnyXu** [comment]
@moxiegirl Sorry I didn't quite understand the procedure in the link you provided. 

And I think I mistakenly pushed the rebased branch to this PR. 

Should I close this PR and recreate it again?

**Comment by moxiegirl** [comment]
Moved into read-to-merge.  This has a small change in hack so I'll leave it for a dev to merge

**Comment by icecrime** [comment]
Hi @TonnyXu! The number of commits in this PR is surprising: could you please rebase and squash appropriately so that the branch is left with only one (signed) additional commit? Thanks!

**Comment by jfrazelle** [comment]
can you please `rebase upstream/master`

**Comment by jfrazelle** [comment]
other maintainer DO NOT MERGE UNTIL COMMITS ARE SQUASHED

**Comment by TonnyXu** [comment]
@jfrazelle @icecrime sorry, I messed this PR when I did a rebase and pushed the rebased branch to this PR. (those commits from others). I will create a new PR and squash the commits

**Comment by icecrime** [comment]
@TonnyXu You don't have to create a new PR, you can simply fix those locally and force push to this branch!

**Comment by thaJeztah** [comment]
@TonnyXu if you need help, feel free to ask on the #docker-dev IRC channel; there's lot of experienced developers willing to give some hints if you need

**Comment by TonnyXu** [comment]
@icecrime @jfrazelle I've pushed the squashed commits, can you confirm?
@thaJeztah Cool! I will check it out.

**Comment by tiborvass** [comment]
@TonnyXu you're all good, thanks for your contribution :)

**Comment by jfrazelle** [comment]
THanks!!

**Comment by jfrazelle** [comment]
LGTM
